### PR TITLE
BAEL-7195 Custom Event Handlers and Listeners in Netty

### PR DESCRIPTION
**Pull request by ulisseslima**
PR recreated due to repository truncation.

Draft: https://drafts.baeldung.com/wp-admin/post.php?post=187064&action=edit

### CODE REVIEW

File: libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/ChatClientMain.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+            future.addListener(new ChannelInfoListener());
+            Channel channel = future.sync()
+                .channel();
+
+            messageLoop(scanner, channel);
+
+            channel.close();
+        } catch (Throwable e) {
+            e.printStackTrace();
```
Comment (JavierPeris) [inline libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/ChatClientMain.java:49]: We should handle the exception rethrowing it or interrupting the thread.
```suggestion
            e.printStackTrace();
            Thread.currentThread().interrupt();
```

File: libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/ChatServerMain.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                .sync();
+
+            System.out.println("chat server started. ready to accept clients.");
+            future.addListener(new ChannelInfoListener());
+
+            future.channel()
+                .closeFuture()
+                .sync();
+        } catch (Throwable e) {
+            e.printStackTrace();
```
Comment (JavierPeris) [inline libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/ChatServerMain.java:46]: Same thing here.

```suggestion
            e.printStackTrace();
            Thread.currentThread().interrupt();
```

File: libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/ChatServerMain.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+public final class ChatServerMain {
+
+    public static final String HOST = "localhost";
+    public static final int PORT = 8081;
+
+    public static void main(String[] args) {
+        EventLoopGroup serverGroup = new NioEventLoopGroup(1);
+        EventLoopGroup clientGroup = new NioEventLoopGroup();
+        try {
```
Comment (JavierPeris) [inline libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/ChatServerMain.java:24]: We can simplify the try with a more updated version, that is, with try-with-resources.
```suggestion
    try (EventLoopGroup clientGroup = new NioEventLoopGroup(); EventLoopGroup serverGroup = new NioEventLoopGroup(1)) {
```

We would remove the finally part too.

File: libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/handler/ServerEventHandler.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Override
+    public void channelActive(final ChannelHandlerContext context) {
+        Channel channel = context.channel();
+        clients.put(channel.id()
+            .asLongText(), channel);
+
+        history.forEach(message -> channel
+            .writeAndFlush(message));
```
Comment (JavierPeris) [inline libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/handler/ServerEventHandler.java:50]: Lambdas should be replaced with method references

```suggestion
        history.forEach(channel::writeAndFlush);
```

File: libraries-server-2/src/test/java/com/baeldung/netty/customhandlersandlisteners/MessageUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        String input = "Bob;Hello, world; go!";
+        Message message = Message.parse(input);
+
+        assertEquals("Bob", message.getUser());
+        assertEquals("Hello, world; go!", message.getMessage());
+        assertNotNull(message.getTime());
+    }
+
+    @Test
+    public void whenNewMessage_thenExpectedFormat() {
```
Comment (JavierPeris) [inline libraries-server-2/src/test/java/com/baeldung/netty/customhandlersandlisteners/MessageUnitTest.java:25]: JUnit5 test classes and methods should have default package visibility.

File: libraries-server-2/src/test/java/com/baeldung/netty/customhandlersandlisteners/ChatIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+import io.netty.channel.embedded.EmbeddedChannel;
+
+class ChatIntegrationTest {
+
+    private static final String MSG_1 = "Alice;Anyone there?!";
+    private static final String MSG_2 = "Bob;Hi, Alice!";
+
+    @Test
+    public void whenMessagesWrittenToServer_thenMessagesConsumed() {
```
Comment (JavierPeris) [inline libraries-server-2/src/test/java/com/baeldung/netty/customhandlersandlisteners/ChatIntegrationTest.java:19]:
```suggestion
    void whenMessagesWrittenToServer_thenMessagesConsumed() {
```

File: libraries-server-2/src/test/java/com/baeldung/netty/customhandlersandlisteners/ChatIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        assertEquals(MSG_1, server.readOutbound()
+            .toString());
+        assertEquals(MSG_2, server.readOutbound()
+            .toString());
+
+        server.close();
+    }
+
+    @Test
+    public void whenClientReceivesMessages_thenMessagesConsumed() {
```
Comment (JavierPeris) [inline libraries-server-2/src/test/java/com/baeldung/netty/customhandlersandlisteners/ChatIntegrationTest.java:37]:
```suggestion
    void whenClientReceivesMessages_thenMessagesConsumed() {
```

File: libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/ChatServerMain.java (hunk 3 above)
Comment (ulisseslima) [reply libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/ChatServerMain.java:24]: This one would result in error "The resource type EventLoopGroup does not implement java.lang.AutoCloseable".

File: libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/listener/ChannelInfoListener.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        Channel channel = future.channel();
+        String status = "OK";
+
+        if (!future.isSuccess()) {
+            status = "FAILED";
+            future.cause()
+                .printStackTrace();
+        }
+
+        System.out.println(String.format("%s - channel#%s %s: %s", Instant.now(), channel.id()
```
Comment (JavierPeris) [inline libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/listener/ChannelInfoListener.java:28]: We can replace println wih printf so we can get rid of String.format

File: libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/listener/ChannelInfoListener.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            .asShortText(), status, event));
```
Comment (JavierPeris) [inline libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/listener/ChannelInfoListener.java:29]:
```suggestion
        System.out.printf("%s - channel#%s %s: %s%n", Instant.now(), channel.id()
            .asShortText(), status, event);
```

File: libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/listener/ChannelInfoListener.java (hunk 8 above)
Comment (ulisseslima) [reply libraries-server-2/src/main/java/com/baeldung/netty/customhandlersandlisteners/listener/ChannelInfoListener.java]: Suggestion included, draft updated.

**Pull request merged**